diffprivlib/accountant.py

```diff
@@ -0,0 +1,294 @@
+            List of tuples of budget spends. If not provided, the accountant's spends will be used.
+
+        slack : float, default : None
+            Slack in delta for composition. If not provided, the accountant's slack will be used.
+
+        Returns
+        -------
+        epsilon : float
+            Total epsilon spend.
+
+        delta : float
+            Total delta spend.
+        """
+        if spent_budget is None:
+            spent_budget = self.__spent_budget
+        else:
+            for epsilon, delta in spent_budget:
+                check_epsilon_delta(epsilon, delta)
+
+        if slack is None:
+            slack = self.slack
+        elif not 0 <= slack <= 1:
+            raise ValueError("Slack must be between 0 and 1 (inclusive), got {}".format(slack))
+
+        epsilon_sum, epsilon_exp_sum, epsilon_sq_sum = 0, 0, 0
+
+        for epsilon, _ in spent_budget:
+            epsilon_sum += epsilon
+            epsilon_exp_sum += (1 - np.exp(-epsilon)) * epsilon / (1 + np.exp(-epsilon))
+            epsilon_sq_sum += epsilon ** 2
+
+        total_epsilon_naive = epsilon_sum
+        total_delta = self.__total_delta_safe(spent_budget, slack)
+
+        if slack == 0:
+            return total_epsilon_naive, total_delta
+
+        total_epsilon_drv = epsilon_exp_sum + np.sqrt(2 * epsilon_sq_sum * np.log(1 / slack))
+        total_epsilon_kov = epsilon_exp_sum + np.sqrt(2 * epsilon_sq_sum *
+                                                      np.log(np.exp(1) + np.sqrt(epsilon_sq_sum) / slack))
+
+        return min(total_epsilon_naive, total_epsilon_drv, total_epsilon_kov), total_delta
+
+    def check_spend(self, epsilon, delta):
+        """Checks if the provided budget can be spent while staying within the accountant's target budget.
+
+        Parameters
+        ----------
+        epsilon : float
+            Epsilon budget spend to check.
+
+        delta : float
+            Delta budget spend to check.
+
+        Returns
+        -------
+        bool
+            True if the budget can be spent, otherwise a :class:`.BudgetError` is raised..
+
+        Raises
+        ------
+        BudgetError
+            If the specified budget spend will result in the target budget being exceeded..
+
+        """
+        check_epsilon_delta(epsilon, delta)
+        spent_budget = self.__spent_budget + [(epsilon, delta)]
+
+        epsilon_spent, delta_spent = self.total_spent(spent_budget=spent_budget)
+
+        if self.epsilon >= epsilon_spent and self.delta >= delta_spent:
+            return True
+
+        raise BudgetError("Privacy spend of ({},{}) not permissible; will exceed remaining privacy budget. "
+                          "Use {}.{} to check remaining budget.".format(epsilon, delta, self.__class__.__name__,
+                                                                        self.remaining_budget.__name__))
+
+    def remaining_budget(self, k=1):
```

Comment:
After some experimentation, this method is quite slow, given the primitive binary search. A more efficient solver might be useful here.